[WIP] Simplify product preferences

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies product preferences form
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Configuration -> Shop Parameters -> Product Preferences. Everything must look the same aside for help now appearing under inputs instead as blue box, also some fields now will appear as required because they always were.

:notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType by PageType and PaginationType. This means if any module extends those types they will get an exception upon upgrading to PS version containing changes in this PR.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Added integer widget as part of the theme

JevgenijVisockij committed:
Making switch not required

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Type/SwitchType.php:
```diff
@@ -52,6 +52,7 @@ public function configureOptions(OptionsResolver $resolver)
             'expanded' => false,
             'disabled' => false,
+            'required' => false,
             'choice_translation_domain' => self::TRANS_DOMAIN,
```
This was already done as part of this PR https://github.com/PrestaShop/PrestaShop/pull/21672 then I accidentaly managed to remove it(probably during the merge or something) as part of this PR https://github.com/PrestaShop/PrestaShop/pull/21245

So here I am adding it again :disappointed:

matks replied:
It happens. No worries. 👍

JevgenijVisockij commented:
>     * button when a product has attributes

I fixed those but does it make sense to Display the "%add_to_cart_label%" button when a product has attributes to be in Admin.Shopparameters.Help when it's a label?

As of "Leave empty to disable." I changed it from "Leave empty to disable" and moved to Admin.Shopparameters.Help as it's part of help now. Is it okay?

JevgenijVisockij committed:
Fixed domains for translations

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -46,29 +46,52 @@ class StockType extends TranslatorAwareType
+                'label' => $this->trans('Delivery time of in-stock products', 'Admin.Shopparameters.Feature'),
+                'help' => $this->trans(
+                        'Advised for European merchants to be legally compliant (eg: Delivered within 3-4 days)',
+                        'Admin.Shopparameters.Help'
```
Should I add dot at the end of help here?

Julievrz replied:
That would be great, thanks!

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -46,29 +46,52 @@ class StockType extends TranslatorAwareType
             ->add('oos_delivery_time', TranslatableType::class, [
+                'label' => $this->trans('Delivery time of out-of-stock products with allowed backorders', 'Admin.Shopparameters.Feature'),
+                'help' => $this->trans('Advised for European merchants to be legally compliant (eg: Delivered within 5-7 days)', 'Admin.Shopparameters.Help'),
                 'type' => TextType::class,
```
Should I add dot at the end of help here?

Julievrz replied:
That would be great, thanks!

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -27,37 +27,73 @@
+            ])
+            ->add('display_unavailable_attributes', SwitchType::class, [
+                'label' => $this->trans('Display unavailable attributes on the product page', 'Admin.Shopparameters.Feature'),
+                'help' => $this->trans('If an attribute is not available in every product combination, it will not be displayed.', 'Admin.Shopparameters.Help'),
```
Is this wording new?

Julievrz replied:
Here too, shouldn't the default status be "No"? Because in the help text we say that if an attribute is not available it will not be displayed but at the same time the button is automatically set on "Yes"

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -46,29 +46,52 @@ class StockType extends TranslatorAwareType
             ])
             ->add('oos_delivery_time', TranslatableType::class, [
+                'label' => $this->trans('Delivery time of out-of-stock products with allowed backorders', 'Admin.Shopparameters.Feature'),
+                'help' => $this->trans('Advised for European merchants to be legally compliant (eg: Delivered within 5-7 days)', 'Admin.Shopparameters.Help'),
```
```suggestion
                'help' => $this->trans('Advised for European merchants to be legally compliant (eg: Delivered within 5-7 days).', 'Admin.Shopparameters.Help'),
```

Julievrz reviewed (commented):
> > ```
> > * button when a product has attributes
> > ```
> 
> I fixed those but does it make sense to Display the "%add_to_cart_label%" button when a product has attributes to be in Admin.Shopparameters.Help when it's a label?
> 
> As of "Leave empty to disable." I changed it from "Leave empty to disable" and moved to Admin.Shopparameters.Help as it's part of help now. Is it okay?

You are right, it is a label so it should be located as a feature. I am currently thinking of adapting the wording here because the label and the tooltip are not complementary. + the feature should be disabled by default and right now it is not. Also, this feature is not available for the classic theme and we're supposed to see a help text "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help (see the specs: https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/shop-parameters/product-settings/product-settings-page.md)

@LouiseBonnard what do you think of the following wording?

Label: "Display the "Add to cart" button when a product has attributes" in Admin.Shopparameters.Feature 
Tooltip: "By default, the "Add to cart" button is hidden when a product has attributes. You can choose to have it displayed in all cases." in Admin.Shopparameters.Help
Help text : "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help

matks reviewed (dismissed):
Code is good 👍  waiting for wording decision

Julievrz commented:
> > > ```
> > > * button when a product has attributes
> > > ```
> > 
> > 
> > I fixed those but does it make sense to Display the "%add_to_cart_label%" button when a product has attributes to be in Admin.Shopparameters.Help when it's a label?
> > As of "Leave empty to disable." I changed it from "Leave empty to disable" and moved to Admin.Shopparameters.Help as it's part of help now. Is it okay?
> 
> You are right, it is a label so it should be located as a feature. I am currently thinking of adapting the wording here because the label and the tooltip are not complementary. + the feature should be disabled by default and right now it is not. Also, this feature is not available for the classic theme and we're supposed to see a help text "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help (see the specs: https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/shop-parameters/product-settings/product-settings-page.md)
> 
> @LouiseBonnard what do you think of the following wording?
> 
> Label: "Display the "Add to cart" button when a product has attributes" in Admin.Shopparameters.Feature
> Tooltip: "By default, the "Add to cart" button is hidden when a product has attributes. You can choose to have it displayed in all cases." in Admin.Shopparameters.Help
> Help text : "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help

So I talked about it with @LouiseBonnard and I think we should just keep the help text:

"Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help"

and get rid of the tooltip (the feature can be specified in the user documentation if needed) 

By default, the feature should be disabled.

JevgenijVisockij committed:
Added dot to the end of the help texts

JevgenijVisockij committed:
Changed label and help for a feature

JevgenijVisockij committed:
Added parameter to labels translation

JevgenijVisockij commented:
> > > > ```
> > > > * button when a product has attributes
> > > > ```
> > > 
> > > 
> > > I fixed those but does it make sense to Display the "%add_to_cart_label%" button when a product has attributes to be in Admin.Shopparameters.Help when it's a label?
> > > As of "Leave empty to disable." I changed it from "Leave empty to disable" and moved to Admin.Shopparameters.Help as it's part of help now. Is it okay?
> > 
> > 
> > You are right, it is a label so it should be located as a feature. I am currently thinking of adapting the wording here because the label and the tooltip are not complementary. + the feature should be disabled by default and right now it is not. Also, this feature is not available for the classic theme and we're supposed to see a help text "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help (see the specs: https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/shop-parameters/product-settings/product-settings-page.md)
> > @LouiseBonnard what do you think of the following wording?
> > Label: "Display the "Add to cart" button when a product has attributes" in Admin.Shopparameters.Feature
> > Tooltip: "By default, the "Add to cart" button is hidden when a product has attributes. You can choose to have it displayed in all cases." in Admin.Shopparameters.Help
> > Help text : "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help
> 
> So I talked about it with @LouiseBonnard and I think we should just keep the help text:
> 
> "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help"
> 
> and get rid of the tooltip (the feature can be specified in the user documentation if needed)
> 
> By default, the feature should be disabled.

Added those changes just to confirm this is what you exepected this setting to look like?
![image](https://user-images.githubusercontent.com/47781757/102899882-4174c980-4474-11eb-8abc-a80447dd23c7.png)

JevgenijVisockij committed:
Added dot to help message

JevgenijVisockij commented on install-dev/data/xml/configuration.xml:
```diff
@@ -228,7 +228,7 @@
   <configuration id="PS_ATTRIBUTE_CATEGORY_DISPLAY" name="PS_ATTRIBUTE_CATEGORY_DISPLAY">
-    <value>1</value>
+    <value>0</value>
   </configuration>
```
allow_add_variant_to_cart_from_listing (PS_ATTRIBUTE_CATEGORY_DISPLAY) should be disabled by default. If I understand correctly the default values are taken from this xml.
https://github.com/PrestaShop/PrestaShop/pull/22471#issuecomment-748831978

Progi1984 replied:
OK with this @PrestaShop/product-team ?

LouiseBonnard replied:
Yep, thanks for asking, @Progi1984! As written in the Product Settings page [specifications](https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/shop-parameters/product-settings/product-settings-page.md#product-page):

> Display the "Add to cart" button when a product has attributes. **Important note: this feature does not work with the Classic default theme.** It has been removed from the front office in 1.7.x because many users added products to their cart without selecting a combination. [...] **It is still in the back office since it might work with other themes. By default, it should be disabled.** [...]

Progi1984 replied:
Ok, so I approve.

Julievrz commented:
> > > > > ```
> > > > > * button when a product has attributes
> > > > > ```
> > > > 
> > > > 
> > > > I fixed those but does it make sense to Display the "%add_to_cart_label%" button when a product has attributes to be in Admin.Shopparameters.Help when it's a label?
> > > > As of "Leave empty to disable." I changed it from "Leave empty to disable" and moved to Admin.Shopparameters.Help as it's part of help now. Is it okay?
> > > 
> > > 
> > > You are right, it is a label so it should be located as a feature. I am currently thinking of adapting the wording here because the label and the tooltip are not complementary. + the feature should be disabled by default and right now it is not. Also, this feature is not available for the classic theme and we're supposed to see a help text "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help (see the specs: https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/shop-parameters/product-settings/product-settings-page.md)
> > > @LouiseBonnard what do you think of the following wording?
> > > Label: "Display the "Add to cart" button when a product has attributes" in Admin.Shopparameters.Feature
> > > Tooltip: "By default, the "Add to cart" button is hidden when a product has attributes. You can choose to have it displayed in all cases." in Admin.Shopparameters.Help
> > > Help text : "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help
> > 
> > 
> > So I talked about it with @LouiseBonnard and I think we should just keep the help text:
> > "Note that this setting does not work with the default theme anymore." in Admin.Shopparameters.Help"
> > and get rid of the tooltip (the feature can be specified in the user documentation if needed)
> > By default, the feature should be disabled.
> 
> Added those changes just to confirm this is what you exepected this setting to look like?
> 
> ![image](https://user-images.githubusercontent.com/47781757/102900009-6701d300-4474-11eb-8381-41634000dd3c.png)

Yes, thank you !

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +48,76 @@ class GeneralType extends TranslatorAwareType
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```
I have been told to favor "integer" instead of "integer number" yesterday so I'll change it when I see it. :) Sorry about that.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +48,76 @@ class GeneralType extends TranslatorAwareType
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +48,76 @@ class GeneralType extends TranslatorAwareType
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +48,76 @@ class GeneralType extends TranslatorAwareType
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -27,37 +27,74 @@
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -27,37 +27,74 @@
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -26,26 +26,47 @@
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -26,26 +26,47 @@
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter a positive integer number.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter a positive integer.', 'Admin.Notifications.Error'),
```

Julievrz reviewed (commented):
The last modification before approving the wording for this PR. Thank you!

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php

Removed number from error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij commented:
> The last modification before approving the wording for this PR. Thank you!

Added those modifications, thanks!

JevgenijVisockij committed:
Made product preferences use translate component

Progi1984 reviewed (commented):
A simple question about the configuration change

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have an issue following these steps:

1. Enable Multishop context in the BO > Shop parameters > General page
2. Create a new shop
3. Go to BO > Shop Parameters > Product Settings page 
4. Select the second shop
5. Enable Catalog mode
6. Select the first shop
7. See error

![image](https://user-images.githubusercontent.com/16067358/104017178-ee12b300-51b7-11eb-8561-16babdd8ad30.png)

It is ok without PR.

Thank you!

JevgenijVisockij commented:
WIP until this is resolved https://github.com/PrestaShop/PrestaShop/pull/21553/files#r554538559

JevgenijVisockij committed:
Simplified product preferences

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig
#	src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_page.html.twig
#	src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_pagination.html.twig
#	src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_stock.html.twig

JevgenijVisockij committed:
Restored product controller back to original

JevgenijVisockij committed:
Changed the way errors are passed to controller and handled there

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/ProductPreferencesController.php:
```diff
@@ -169,18 +175,82 @@ protected function processForm(Request $request, FormHandlerInterface $formHandl
+        switch ($error->getErrorCode()) {
+            case GeneralFormDataProvider::ERROR_MUST_BE_NUMERIC_EQUAL_TO_ZERO_OR_HIGHER:
+                return $this->trans(
+                    '%s is invalid. Please enter an integer greater or equal to 0.',
```
```suggestion
                    '%s is invalid. Please enter an integer greater than or equal to 0.',
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/ProductPreferencesController.php

Add then to a translation

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralFormDataProvider.php:
```diff
@@ -37,6 +41,10 @@
  */
 class GeneralFormDataProvider implements FormDataProviderInterface
 {
+    /** @todo change error to string once invoices PR merged  */
```
Which PR is involved ?

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -27,37 +27,82 @@
-class PageType extends AbstractType
+class PageType extends TranslatorAwareType
 {
+    public const FIELD_DISPLAY_QUANTITIES = 'display_quantities';
```
Why do you put all field names in constant ?

Progi1984 commented:
@JevgenijVisockij Please fix PHPCSFixer

PierreRambaud commented:
Hey @JevgenijVisockij 
Are you still working on it? Should we close this one?
If no one is working on it, maybe someone from the community is able to finish the job in a new PR?

Kind regards
